rclpy/rclpy/action/client.py

```diff
@@ -437,23 +447,23 @@ def send_goal_async(self, goal, feedback_callback=None, goal_uuid=None):
         request = self._action_type.Impl.SendGoalService.Request()
         request.goal_id = self._generate_random_uuid() if goal_uuid is None else goal_uuid
         request.goal = goal
-        sequence_number = self._client_handle.send_goal_request(request)
-        if sequence_number in self._pending_goal_requests:
-            raise RuntimeError(
-                'Sequence ({}) conflicts with pending goal request'.format(sequence_number))
+        future = Future()
+        with self._lock:
+            sequence_number = self._client_handle.send_goal_request(request)
+            if sequence_number in self._pending_goal_requests:
+                raise RuntimeError(
+                    'Sequence ({}) conflicts with pending goal request'.format(sequence_number))
+            self._pending_goal_requests[sequence_number] = future
+            self._goal_sequence_number_to_goal_id[sequence_number] = request.goal_id
+            future.add_done_callback(self._remove_pending_goal_request)
+            # Add future so executor is aware
+            self.add_future(future)
 
         if feedback_callback is not None:
             # TODO(jacobperron): Move conversion function to a general-use package
             goal_uuid = bytes(request.goal_id.uuid)
             self._feedback_callbacks[goal_uuid] = feedback_callback
```

Comment:
Brainstorming a bit here. Should the feedback callback block also be brought into the lock, since it edits the `self._feedback_callbacks`, which could at the moment be accessed by concurrently once the `self._pending_goal_requests` and `self._goal_sequence_number_to_goal_id` dicts have been added?

```suggestion
        future = Future()
        with self._lock:
            sequence_number = self._client_handle.send_goal_request(request)
            if sequence_number in self._pending_goal_requests:
                raise RuntimeError(
                    'Sequence ({}) conflicts with pending goal request'.format(sequence_number))
            self._pending_goal_requests[sequence_number] = future
            self._goal_sequence_number_to_goal_id[sequence_number] = request.goal_id
            future.add_done_callback(self._remove_pending_goal_request)
            # Add future so executor is aware
            self.add_future(future)

            if feedback_callback is not None:
                # TODO(jacobperron): Move conversion function to a general-use package
                goal_uuid = bytes(request.goal_id.uuid)
                self._feedback_callbacks[goal_uuid] = feedback_callback
```

Reply:
With the current implementation it looks like we only need to lock calls to non-python types i.e self._client_handle, since python is *mostly* thread safe. It may be better to lock more things but this appears to work, if we want to put locks on _feedback_callbacks we'll need to consider a more major refactor to consider all the other global dictionaries as well.